drv/lpc55-update-server/src/images.rs

```diff
@@ -23,82 +31,605 @@ extern "C" {
+    }
+    Err(RawCabooseError::MissingCaboose)
+}
 
-    let reset_vector = u32::from_le_bytes(
-        block[RESET_VECTOR_OFFSET..][..4].try_into().unwrap_lite(),
-    ) & ADDRMASK;
+/// Perform sanity checking on the header block and retrieve
+/// The Epoch from the optional ImageHeader and the offset to
+/// the end of the image/beginning of the signature block.
+///
+/// Return an Epoch if available and the length of the image.
+/// Note the length of the executable image is also the offset to the beginning
+/// signature block and the end of the caboose.
+pub fn validate_header_block(
+    header_access: &ImageAccess<'_>,
+    component: RotComponent,
+    slot: SlotId,
+) -> Result<(Option<Epoch>, usize), UpdateError> {
+    // let mut vectors: &mut [u8] = &mut [0u32; BLOCK_SIZE_BYTES / SIZEOF_U32].as_bytes();
+    let mut vectors = ImageVectorsLpc55::new_zeroed();
+    let mut header = ImageHeader::new_zeroed();
 
-    // Ensure the image is destined for the right target
-    if !exec.contains(&reset_vector) {
+    if header_access
+        .read_bytes(
+            0,
+            core::mem::size_of::<ImageVectorsLpc55>() as u32,
+            vectors.as_bytes_mut(),
+        )
+        .is_err()
+        || header_access
+            .read_bytes(
+                IMAGE_HEADER_OFFSET,
+                core::mem::size_of::<ImageHeader>() as u32,
+                header.as_bytes_mut(),
+            )
+            .is_err()
+    {
+        // can't read block0
         return Err(UpdateError::InvalidHeaderBlock);
     }
 
-    // Ensure the MAGIC is correct.
-    // Bootloaders have been released without an ImageHeader. Allow those.
-    let magic =
-        u32::from_le_bytes(block[MAGIC_OFFSET..][..4].try_into().unwrap_lite());
-    if component == RotComponent::Hubris && magic != abi::HEADER_MAGIC {
+    // Ensure the image is destined for the right target
+    let exec = &flash_range(component, slot).exec;
+    if !vectors.is_image_type_signed() {
         return Err(UpdateError::InvalidHeaderBlock);
     }
+    let padded_range =
+        if let Some(padded_range) = vectors.padded_image_range(exec) {
+            padded_range
+        } else {
+            return Err(UpdateError::InvalidHeaderBlock);
+        };
 
-    Ok(())
+    // Get the epoch if it is available.
+    let header_epoch = if header.magic == HEADER_MAGIC {
+        if header.total_image_len != vectors.nxp_offset_to_specific_header {
+            return Err(UpdateError::InvalidHeaderBlock);
+        }
+        Some(Epoch::from(header.epoch))
+    } else {
+        None
+    };
+
+    Ok((header_epoch, padded_range.len()))
 }
 
 pub fn same_image(component: RotComponent, slot: SlotId) -> bool {
     // Safety: We are trusting the linker.
-    image_range(component, slot).0.start
+    flash_range(component, slot).store.start
         == unsafe { &__this_image } as *const _ as u32
 }
 
+pub struct FlashRange {
+    pub store: Range<u32>,
+    pub exec: Range<u32>,
+}
+
```

Comment:
Can we break this part of the change out into a separate PR to review?

Reply:
Done PR #1813